lib/backend/pgbk/pgbk.go

```diff
@@ -451,61 +480,77 @@ func (b *Backend) Get(ctx context.Context, key []byte) (*backend.Item, error) {
 
 // GetRange implements [backend.Backend].
 func (b *Backend) GetRange(ctx context.Context, startKey []byte, endKey []byte, limit int) (*backend.GetResult, error) {
-	if limit <= 0 {
-		limit = backend.DefaultRangeLimit
-	}
-
-	items, err := pgcommon.RetryIdempotent(ctx, b.log, func() ([]backend.Item, error) {
-		batch := new(pgx.Batch)
-		// batches run in an implicit transaction
-		batch.Queue("SET transaction_read_only TO on")
-		// TODO(espadolini): figure out if we want transaction_deferred enabled
-		// for GetRange
-
-		var items []backend.Item
-		batch.Queue(
-			"SELECT kv.key, kv.value, kv.expires, kv.revision FROM kv"+
-				" WHERE kv.key BETWEEN $1 AND $2 AND (kv.expires IS NULL OR kv.expires > now())"+
-				" ORDER BY kv.key LIMIT $3",
-			nonNil(startKey), nonNil(endKey), limit,
-		).Query(func(rows pgx.Rows) error {
-			var err error
-			items, err = pgx.CollectRows(rows, func(row pgx.CollectableRow) (backend.Item, error) {
-				var key, value []byte
-				var expires time.Time
-				var revision revision
-				if err := row.Scan(&key, &value, (*zeronull.Timestamptz)(&expires), &revision); err != nil {
-					return backend.Item{}, err
-				}
-				return backend.Item{
-					Key:      key,
-					Value:    value,
-					Expires:  expires.UTC(),
-					ID:       idFromRevision(revision),
-					Revision: revisionToString(revision),
-				}, nil
+	lastSeenKey := startKey
+	results := &backend.GetResult{}
+	pageSize := defaultPageSize
+
+	if limit > 0 && pageSize > limit {
+		pageSize = limit
+	}
+
+	for {
+		items, err := pgcommon.RetryIdempotent(ctx, b.log, func() ([]backend.Item, error) {
+			batch := new(pgx.Batch)
+			// batches run in an implicit transaction
+			batch.Queue("SET transaction_read_only TO on")
+			// TODO(espadolini): figure out if we want transaction_deferred enabled
+			// for GetRange
+
+			var items []backend.Item
+			batch.Queue(
+				"SELECT kv.key, kv.value, kv.expires, kv.revision FROM kv"+
+					" WHERE kv.key BETWEEN $1 AND $2 AND (kv.expires IS NULL OR kv.expires > now())"+
+					" ORDER BY kv.key LIMIT $3",
+				NonNil(lastSeenKey), NonNil(endKey), pageSize,
+			).Query(func(rows pgx.Rows) error {
```

Comment:
There's no need to keep track of a changing `pageSize`, it just adds confusion.
```suggestion
			var items []backend.Item
			stepLimit := min(len(results.Items) - limit, defaultPageSize)
			batch.Queue(
				"SELECT kv.key, kv.value, kv.expires, kv.revision FROM kv"+
					" WHERE kv.key BETWEEN $1 AND $2 AND (kv.expires IS NULL OR kv.expires > now())"+
					" ORDER BY kv.key LIMIT $3",
				NonNil(lastSeenKey), NonNil(endKey), stepLimit,
			).Query(func(rows pgx.Rows) error {
```
After this, we know that we ought to stop iterating if `len(items) < stepLimit`.